### PR TITLE
WIP feat: add mapState to digi pack landing page

### PR DESCRIPTION
**Pull request by damien-shaw-guardian**
After launching the digital pack landing page (which was an A/B test between the first and second screenshot), we noticed that sometimes people would be shown the original page, while being identified to Optimize (the A/B testing platform) as being part of the Variant sample, this meant the results for the test are invalid as we could not be sure which view the user was seeing 

## Why are you doing this?

[**Trello Card**](https://trello.com)

## Changes

- add mapState to digital pack landing page, to allow the page to re-render when there is a change to the redux store
- change LandingPage component to a class Component, in order to as state to facilitate a re-render
- added an anti flicker function to control when the page is shown
- removed a call to optimize.js, so optimize.js it only called once 
- add polling to optimize.js to determine when the Google dataLayer has loaded

**Original**
Screen Shot 2019-06-10 at 16 32 06

**New Variant**
Screen Shot 2019-06-26 at 17 26 56

### CODE REVIEW

File: support-frontend/package.json [hunk 1]
```diff
@@ -126,6 +126,7 @@
     "rimraf": "^2.6.2",
     "sass-loader": "^7.1.0",
     "sass-mq": "^5.0.0",
+    "uglifyjs-webpack-plugin": "^2.1.3",
```
Comment (rupertbates) [inline support-frontend/package.json]: What is this needed for?

Comment (damien-shaw-guardian) [reply support-frontend/package.json]: this was added to allow vscode to read the import paths properly, I could add this to a different PR as that better for tracking

**Comment by jfsoul** [comment]
Would you mind updating the description with an explanation of the steps we're taking in this PR and why?

File: support-frontend/assets/pages/digital-subscription-landing/digitalSubscriptionLanding.jsx [hunk 2]
```diff
@@ -69,56 +68,119 @@ const CountrySwitcherHeader = headerWithCountrySwitcherContainer({
...
+  state = {
+    showPage: this.props.optimizeHasLoaded,
+    numOfChecks: 0,
+  }
+
+  wait = () => {
+    const { numOfChecks } = this.state;
+    const { optimizeHasLoaded } = this.props;
+
+    if (optimizeHasLoaded || numOfChecks > 16) {
```
Comment (jfsoul) [inline support-frontend/assets/pages/digital-subscription-landing/digitalSubscriptionLanding.jsx]: Is this now just to deal with the case where optimize never loads within the other timeout? I assume the case where optimize does eventually load is handled just by updating the state.

If that's right then can't we just update another bit of state (`showPage`) from `optimize.js` in the path where the timeout is reached?

File: support-frontend/assets/helpers/optimize/optimize.js [hunk 3]
```diff
@@ -103,22 +141,12 @@ function addOptimizeExperiments(addToStoreCallback: (OptimizeExperiment) => void
   if (optimizeIsLoaded()) {
     getExperimentsFromApi((variant, id) => withSessionStorageCallback({ id, variant }));
   } else {
-    // Add a listener so we can update the store once Optimize loads
+  // Add a listener so we can update the store once Optimize loads
     addExperimentUpdateListener(withSessionStorageCallback);
+    getOptimizeExperiments();
```
[the remark this answers is not in the document]
Comment (damien-shaw-guardian) [reply support-frontend/assets/helpers/optimize/optimize.js]: we need this call, as this path calls gtag to begin polling for the dataLayer

File: support-frontend/assets/pages/digital-subscription-landing/components/theMoment.scss [hunk 4]
```diff
@@ -273,6 +273,6 @@
 
   @include mq($from: tablet) {
     font-size: 24px;
-    line-height: 21px;
+    line-height: 28px;
```
[the remark this answers is not in the document]
Comment (damien-shaw-guardian) [reply support-frontend/assets/pages/digital-subscription-landing/components/theMoment.scss]: this was a quick fix, Gus asked for

File: support-frontend/assets/pages/digital-subscription-landing/digitalSubscriptionLanding.jsx [hunk 5]
```diff
@@ -69,56 +68,119 @@ const CountrySwitcherHeader = headerWithCountrySwitcherContainer({
   ],
 });
 
-const { optimizeExperiments } = store.getState().common;
+const mapStateToProps = (state) => {
+  const { optimizeExperiments } = state.common;
+  const dailyEditionsExperimentId = 'ZVlxqx3wRDGSDw-HYZLiAQ';
```
[the remark this answers is not in the document]
Comment (damien-shaw-guardian) [reply support-frontend/assets/pages/digital-subscription-landing/digitalSubscriptionLanding.jsx]: good catch, yes we will need to make a new experiment too

File: support-frontend/assets/helpers/optimize/optimize.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jfsoul) [reply support-frontend/assets/helpers/optimize/optimize.js]: then do we need the listener any longer?

File: support-frontend/assets/pages/digital-subscription-landing/digitalSubscriptionLanding.jsx [hunk 6]
```diff
@@ -69,56 +68,119 @@ const CountrySwitcherHeader = headerWithCountrySwitcherContainer({
...
+  const dailyEditionsVariant2 = optimizeExperiments
+    .filter(exp => exp.id === dailyEditionsExperimentId && exp.variant === '1').length !== 0;
+
+  return {
+    dailyEditionsVariant: dailyEditionsVariant2 != null ? dailyEditionsVariant2 : false,
```
[the remark this answers is not in the document]
Comment (damien-shaw-guardian) [reply support-frontend/assets/pages/digital-subscription-landing/digitalSubscriptionLanding.jsx]: this was a problem for flow at one point, but it seems to be ok now

File: support-frontend/assets/pages/digital-subscription-landing/digitalSubscriptionLanding.jsx (hunk 2 above)
Comment (damien-shaw-guardian) [reply support-frontend/assets/pages/digital-subscription-landing/digitalSubscriptionLanding.jsx]: once the component is initialised, state can only be updated by setState (as far as I know) so I don't think we can update it from optimize.js, also I think it's better to have presentation logic controlled by the component

File: support-frontend/assets/helpers/optimize/optimize.js [hunk 7]
```diff
@@ -72,9 +96,12 @@ function storeExperimentInSession(experiment: OptimizeExperiment): boolean {
 }
 
 function addExperimentUpdateListener(addToStoreCallback: OptimizeExperiment => void) {
+  console.log('addExperimentUpdateListener');
   window.addEventListener(
     EXPERIMENTS_UPDATED,
-    ev => addToStoreCallback({ id: ev.detail.id, variant: ev.detail.variant }),
+    (ev) => {
+      addToStoreCallback({ id: ev.detail.id, variant: ev.detail.variant });
+    },
```
[the remark this answers is not in the document]
Comment (damien-shaw-guardian) [reply support-frontend/assets/helpers/optimize/optimize.js]: i just returned to this to the original format

File: support-frontend/assets/helpers/optimize/optimize.js [hunk 8]
```diff
@@ -27,13 +27,37 @@ function optimizeIsLoaded() {
...
+  //   setTimeout(() => {
+  //     // eslint-disable-next-line prefer-spread
+  //     gtag.apply(null, arguments); // eslint-disable-line prefer-rest-params
+  //   }, 250);
+  // }
 }
 
 function getExperimentsFromApi(callback: (variant: string, id: string) => void) {
+  // the Timeout is to delay the callback until the eventListener has been added to the DOM
+  console.log('callback', callback);
```
[the remark this answers is not in the document]
Comment (damien-shaw-guardian) [reply support-frontend/assets/helpers/optimize/optimize.js]: Removed

File: support-frontend/assets/helpers/optimize/optimize.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (damien-shaw-guardian) [reply support-frontend/assets/helpers/optimize/optimize.js]: we need both, if we want to have an instant update when the dataLayer is all ready available and a callback when the dataLayer is not, alternatively we could just have  a polling callback and we remove the instant update which I imagine would be slower but not by much, maybe 200ms I would guess

File: support-frontend/assets/pages/digital-subscription-landing/digitalSubscriptionLanding.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (damien-shaw-guardian) [reply support-frontend/assets/pages/digital-subscription-landing/digitalSubscriptionLanding.jsx]: I removed dailyEditionsVariant2

File: support-frontend/assets/pages/digital-subscription-landing/digitalSubscriptionLanding.jsx [hunk 9]
```diff
@@ -81,56 +72,118 @@ const CountrySwitcherHeader = headerWithCountrySwitcherContainer({
...
+      this.setState(() => ({
+        showPage: true,
+      }));
+    } else {
+      this.setState(prevState => ({
+        pageReadyChecks: prevState.pageReadyChecks + 1,
+      }));
+
+      setTimeout(() => {
+        this.checkOptimizeIsReady(250);
```
Comment (jfsoul) [inline support-frontend/assets/pages/digital-subscription-landing/digitalSubscriptionLanding.jsx]:
```suggestion
        this.checkOptimizeIsReady(interval);
```

**Pull request merged**